File: Week 9/Source Code/controllers/camera_recognition/camera_recognition.c
#include <stdio.h>
#include <webots/camera.h>
#include <webots/camera_recognition_object.h>
#include <webots/motor.h>
#include <webots/robot.h>

// Mendefinisikan kecepatan motor dan langkah waktu simulasi
#define SPEED 1.5
#define TIME_STEP 64

int main() {
  // Deklarasi tag perangkat untuk kamera, motor kiri, dan motor kanan
  WbDeviceTag camera, left_motor, right_motor;
  int i, j; // Variabel untuk iterasi

  // Inisialisasi robot
  wb_robot_init();

  /* Mengambil perangkat kamera, mengaktifkan kamera, dan mengaktifkan fitur pengenalan objek */
  camera = wb_robot_get_device("camera");
  wb_camera_enable(camera, TIME_STEP);
  wb_camera_recognition_enable(camera, TIME_STEP);

  /* Mengambil handler untuk motor kiri dan kanan, 
     serta mengatur posisi target motor ke infinity (kontrol kecepatan) */
  left_motor = wb_robot_get_device("left wheel motor");
  right_motor = wb_robot_get_device("right wheel motor");
  wb_motor_set_position(left_motor, INFINITY);
  wb_motor_set_position(right_motor, INFINITY);

  /* Mengatur kecepatan awal motor:
     - Motor kiri bergerak mundur dengan kecepatan -SPEED
     - Motor kanan bergerak maju dengan kecepatan SPEED */
  wb_motor_set_velocity(left_motor, -SPEED);
  wb_motor_set_velocity(right_motor, SPEED);

  /* Loop utama simulasi */
  while (wb_robot_step(TIME_STEP) != -1) {
    /* Mendapatkan jumlah objek yang dikenali oleh kamera */
    int number_of_objects = wb_camera_recognition_get_number_of_objects(camera);
    printf("\nRecognized %d objects.\n", number_of_objects);

    /* Mengambil informasi objek yang dikenali dan menampilkannya */
    const WbCameraRecognitionObject *objects = wb_camera_recognition_get_objects(camera);
    for (i = 0; i < number_of_objects; ++i) {
      // Menampilkan model objek
      printf("Model of object %d: %s\n", i, objects[i].model);
      
      // Menampilkan ID objek
      printf("Id of object %d: %d\n", i, objects[i].id);
      
      // Menampilkan posisi relatif objek
      printf("Relative position of object %d: %lf %lf %lf\n", i, objects[i].position[0], objects[i].position[1],
             objects[i].position[2]);
      
      // Menampilkan orientasi relatif objek
      printf("Relative orientation of object %d: %lf %lf %lf %lf\n", i, objects[i].orientation[0], objects[i].orientation[1],
             objects[i].orientation[2], objects[i].orientation[3]);
      
      // Menampilkan ukuran objek
      printf("Size of object %d: %lf %lf\n", i, objects[i].size[0], objects[i].size[1]);
      
      // Menampilkan posisi objek pada gambar kamera
      printf("Position of the object %d on the camera image: %d %d\n", i, objects[i].position_on_image[0],
             objects[i].position_on_image[1]);
      
      // Menampilkan ukuran objek pada gambar kamera
      printf("Size of the object %d on the camera image: %d %d\n", i, objects[i].size_on_image[0], objects[i].size_on_image[1]);
      
      // Menampilkan warna-warna yang dikenali pada objek
      for (j = 0; j < objects[i].number_of_colors; ++j)
        printf("- Color %d/%d: %lf %lf %lf\n", j + 1, objects[i].number_of_colors, objects[i].colors[3 * j],
               objects[i].colors[3 * j + 1], objects[i].colors[3 * j + 2]);
    }
  }

  // Membersihkan alokasi sumber daya robot setelah simulasi selesai
  wb_robot_cleanup();

  return 0;
}

File: Week 9/Source Code/controllers/camera_recognition/camera_recognition.py
from controller import Robot

class Controller(Robot):
    def __init__(self):
        # Memanggil konstruktor kelas induk Robot
        super(Controller, self).__init__()
        
        # Mendefinisikan langkah waktu untuk simulasi
        self.timeStep = 64
        
        # Mengambil perangkat kamera dan mengaktifkannya
        self.camera = self.getDevice('camera')
        self.camera.enable(self.timeStep)
        
        # Mengaktifkan fitur pengenalan objek pada kamera
        self.camera.recognitionEnable(self.timeStep)
        
        # Mengambil perangkat motor kiri dan kanan
        self.left_motor = self.getDevice('left wheel motor')
        self.right_motor = self.getDevice('right wheel motor')
        
        # Mengatur posisi motor ke infinity untuk kontrol kecepatan
        self.left_motor.setPosition(float('inf'))
        self.right_motor.setPosition(float('inf'))
        
        # Mengatur kecepatan awal motor
        # Motor kiri bergerak mundur (-1.5), motor kanan bergerak maju (1.5)
        self.left_motor.setVelocity(-1.5)
        self.right_motor.setVelocity(1.5)

    def run(self):
        # Loop utama untuk menjalankan simulasi
        while self.step(self.timeStep) != -1:
            # Mendapatkan jumlah objek yang dikenali oleh kamera
            number_of_objects = self.camera.getRecognitionNumberOfObjects()
            print(f'Recognized {number_of_objects} objects.')
            print(' ')
            
            # Mengambil informasi semua objek yang dikenali
            objects = self.camera.getRecognitionObjects()
            counter = 1  # Inisialisasi penghitung objek
            
            for object in objects:
                # Mendapatkan atribut objek: posisi, orientasi, ukuran, dll.
                position = object.getPosition()  # Posisi relatif objek
                orientation = object.getOrientation()  # Orientasi relatif objek
                size = object.getSize()  # Ukuran objek
                position_on_image = object.getPositionOnImage()  # Posisi pada gambar kamera
                size_on_image = object.getSizeOnImage()  # Ukuran pada gambar kamera
                number_of_colors = object.getNumberOfColors()  # Jumlah warna yang dikenali
                colors = object.getColors()  # Warna objek
                
                # Menampilkan informasi lengkap tentang objek
                print(f' Object {counter}/{number_of_objects}: {object.getModel()} (id = {object.getId()})')
                print(f' Position: {position[0]} {position[1]} {position[2]}')
                print(f' Orientation: {orientation[0]} {orientation[1]} {orientation[2]} {orientation[3]}')
                print(f' Size: {size[0]} x {size[1]}')
                print(f' Position on camera image: {position_on_image[0]} {position_on_image[1]}')
                print(f' Size on camera image: {size_on_image[0]} x {size_on_image[1]}')
                
                # Menampilkan semua warna yang dikenali pada objek
                for j in range(number_of_colors):
                    print(f'  Color {j + 1}/{number_of_colors}: '
                          f'{colors[3 * j]} {colors[3 * j + 1]} {colors[3 * j + 2]}')
                print(' ')
                
                # Meningkatkan penghitung objek
                counter += 1

# Membuat instance dari kelas Controller dan menjalankan simulasi
controller = Controller()
controller.run()
